### PR TITLE
Feature/communication server calling v3

### PR DESCRIPTION
**Pull request by minwoolee-msft**
Initial Calling Server V3 merge from private repo

### CODE REVIEW

**Comment by yyw-msft** [comment]
do we need to update the README.md?

**Comment by minwoolee-msft** [comment]
> do we need to update the README.md?

Updated README to have latest codes

**Comment by yyw-msft** [comment]
I dont have an answer to this but not sure if we should remove all the existing tests from v2?

**Comment by juntuchen-msft** [comment]
> I dont have an answer to this but not sure if we should remove all the existing tests from v2?

I deleted them in Java repo though.

**Comment by cochi2** [comment]
@yyw-msft the only tests that are here are for Download/Delete recording, which didn't change.

**Comment by minwoolee-msft** [comment]
> I dont have an answer to this but not sure if we should remove all the existing tests from v2?

Test needs to be re-written as most of API and SDK methods are all new, and re-recorded
This can be separate PR or merged into this if it is ready

File: sdk/communication/Azure.Communication.CallingServer/src/ContentCapabilities.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        /// <param name="playSource"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<Response<PlayResponse>> PlayToAllAsync(PlaySource playSource, CancellationToken cancellationToken = default)
+        {
+            return await PlayAsync(playSource, Enumerable.Empty<CommunicationIdentifier>(), cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Plays a file async.
```
Comment (jasonshave) [inline sdk/communication/Azure.Communication.CallingServer/src/ContentCapabilities.cs:51]: [nit] this is a non-async call

File: sdk/communication/Azure.Communication.CallingServer/src/Models/AddParticipantsOptions.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    public class AddParticipantsOptions
+    {
+        /// <summary> The alternate caller id of the source. </summary>
+        public PhoneNumberIdentifier AlternateCallerId { get; set; }
+
+        /// <summary> The Operation Context. </summary>
+        public string OperationContext { get; set; }
+
+        /// <summary> Timeout before invitation timesout. </summary>
+        public int invitationTimeoutInSeconds { get; set; }
```
Comment (jasonshave) [inline sdk/communication/Azure.Communication.CallingServer/src/Models/AddParticipantsOptions.cs:21]: [nit] casing

File: sdk/communication/Azure.Communication.CallingServer/src/Models/AddParticipantsOptions.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+        /// <summary> The replacement CallConnectionId </summary>
+        public string replacementCallConnectionId { get; set; }
```
Comment (jasonshave) [inline sdk/communication/Azure.Communication.CallingServer/src/Models/AddParticipantsOptions.cs:24]: [nit] casing

File: sdk/communication/Azure.Communication.CallingServer/src/Models/CallConnectionProperties.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            /*
+            CallLegId = callConnectionPropertiesInternal.CallLegId;
+            Source = CommunicationIdentifierSerializer.Deserialize(callConnectionPropertiesInternal.Source);
+            Target = CommunicationIdentifierSerializer.Deserialize(callConnectionPropertiesInternal.Target);
+            AlternateCallerId = callConnectionPropertiesInternal.AlternateCallerId == null ? null : new PhoneNumberIdentifier(callConnectionPropertiesInternal.AlternateCallerId.Value);
+            CallConnectionState = callConnectionPropertiesInternal.CallConnectionState;
+            Subject = callConnectionPropertiesInternal.Subject;
+            CallbackUri = new Uri(callConnectionPropertiesInternal.CallbackUri);
+            */
+            var TODO = callConnectionPropertiesInternal;
```
Comment (jasonshave) [inline sdk/communication/Azure.Communication.CallingServer/src/Models/CallConnectionProperties.cs:24]: [nit] todo?

File: sdk/communication/Azure.Communication.CallingServer/src/Models/CallConnectionProperties.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+namespace Azure.Communication.CallingServer
+{
+    /// <summary> The Details of the call. </summary>
+    public class CallConnectionProperties
+    {
+        /// <summary> Initializes a new instance of CallConnectionProperties. </summary>
+        /// <param name="callConnectionPropertiesInternal">The call connection properties internal.</param>
+        internal CallConnectionProperties(CallConnectionPropertiesInternal callConnectionPropertiesInternal)
+        {
+            /*
```
Comment (jasonshave) [inline sdk/communication/Azure.Communication.CallingServer/src/Models/CallConnectionProperties.cs:15]: Should we leave this in here commented out or remove it?

**Review by jasonshave — CHANGES_REQUESTED** [review]
@minwoolee-msft Do we want to put our public event models in from the API spec or add them later?

File: sdk/communication/Azure.Communication.CallingServer/src/CallingServerClient.cs [hunk 6]
```diff
@@ -220,37 +318,44 @@ public virtual async Task<Response<CallConnection>> JoinCallAsync(string serverC
...
         /// <param name="cancellationToken"> The cancellation token. </param>
         /// <exception cref="RequestFailedException">The server returned an error. See <see cref="Exception.Message"/> for details returned from the server.</exception>
         /// <exception cref="ArgumentNullException"><paramref name="source"/> is null.</exception>
-        /// <exception cref="ArgumentNullException"><paramref name="serverCallId"/> is null.</exception>
-        /// <exception cref="ArgumentNullException"><paramref name="callOptions"/> is null.</exception>
-        public virtual Response<CallConnection> JoinCall(string serverCallId, CommunicationIdentifier source, JoinCallOptions callOptions, CancellationToken cancellationToken = default)
+        /// <exception cref="ArgumentNullException"><paramref name="targets"/> is null.</exception>
+        /// <exception cref="ArgumentNullException"><paramref name="options"/> is null.</exception>
+        public virtual Response<CallConnection> CreateCall(CommunicationIdentifier source, IEnumerable<CommunicationIdentifier> targets, Uri callbackUri, CreateCallOptions options = default, CancellationToken cancellationToken = default)
```
Comment (williamzhao87) [inline sdk/communication/Azure.Communication.CallingServer/src/CallingServerClient.cs:331]: Why don't we accept a CallSource object in the method parameters?

File: sdk/communication/Azure.Communication.CallingServer/src/CallConnection.cs [hunk 7]
```diff
@@ -22,11 +24,38 @@ public class CallConnection
...
         public virtual string CallConnectionId { get; internal set; }
 
-        internal CallConnection(string callConnectionId, CallConnectionsRestClient callConnectionRestClient, ClientDiagnostics clientDiagnostics)
+        /// <summary>
+        /// The ServerCall id.
+        /// </summary>
+        public virtual string ServerCallId { get; internal set; }
+
+        /// <summary> The source of the call. </summary>
+        public virtual CommunicationIdentifier Source { get; internal set; }
```
Comment (williamzhao87) [inline sdk/communication/Azure.Communication.CallingServer/src/CallConnection.cs:33]: Should Source and AlternateCallerId be wrapped into a CallSource object here?

File: sdk/communication/Azure.Communication.CallingServer/src/ContentCapabilities.cs (hunk 1 above)
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallingServer/src/ContentCapabilities.cs]: Updated

File: sdk/communication/Azure.Communication.CallingServer/src/Models/CallConnectionProperties.cs (hunk 4 above)
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallingServer/src/Models/CallConnectionProperties.cs]: This is reminiscence of previous work I forgot to delete. Removed.

File: sdk/communication/Azure.Communication.CallingServer/src/Models/CallConnectionProperties.cs (hunk 5 above)
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallingServer/src/Models/CallConnectionProperties.cs]: Removed

File: sdk/communication/Azure.Communication.CallingServer/src/CallingServerClient.cs (hunk 6 above)
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallingServer/src/CallingServerClient.cs]: Good point, created a new object CallSource

File: sdk/communication/Azure.Communication.CallingServer/src/CallConnection.cs (hunk 7 above)
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallingServer/src/CallConnection.cs:33]: Updated

**Review by jasonshave — APPROVED** [review]

**Pull request merged**